### ISSUE TITLE
Long Chapter Titles Cropped on Small Screens When Font Size Too Large

### ISSUE THREAD

**Issue by jdempcy** [readium/readium-js-viewer#474]
Issue:
    - On a small screen, when the font-size is increased beyond a certain amount, some words can be cropped.

![screen shot 2016-01-26 at 2 31 50 pm](https://cloud.githubusercontent.com/assets/95693/12597339/8e5264c6-c439-11e5-9f39-2d6947b3c272.png)

![screen shot 2016-01-26 at 2 32 52 pm](https://cloud.githubusercontent.com/assets/95693/12597373/b86f3270-c439-11e5-97a2-234668ecb525.png)
![screen shot 2016-01-26 at 2 32 57 pm](https://cloud.githubusercontent.com/assets/95693/12597374/b87168c4-c439-11e5-8c5b-5d2ba36b1acc.png)

Note that the latter screenshot is not showing cropped text. It is simply showing how the rest of the (non-header) text renders. I believe that it is reasonable that a user may want to further increase the non-header text. Right now this is not possible without header text becoming larger and thus getting cropped.

Steps to repro:
    - In Chrome or the browser of your choice, navigate to: <https://readium.firebaseapp.com/?epub=https%3A%2F%2Freadium.firebaseapp.com%2Fepub_content%2Fmoby_dick&>.
    - Open Settings and increase the font-size to maximum.
    - Decrease the browser width to minimum.
    - Notice that the word "ETYMOLOGY" now says "ETYMOLO"—the last two characters have been cropped. (On my screen, a few pixels of the edge of the "O" are cropped as well.)

Notes: 

This is more of an issue for devices with small screens. I keep running into cases where long words in chapter titles are cropped when the font-size setting is around 180%. This differs depending on the word and original font size specified by the author.

Would it be possible to restrict the font-size to a maximum size which it cannot exceed, at the time of applying font-setting updates?

I imagine we could calculate the length of the longest word in the element, and then figure out an upper limit on the font size from that.

For instance, if the longest word in the element is 'ETYMOLOGY' with 9 characters, then we would measure how much space we have to work with, say, 300px, and decide that we cannot exceed a font-size beyond which it renders using more than 300 / 9 pixels width, or 33 pixels wide, rounding down.

Then, we could use a, hopefully, "good enough" solution which errs on the side of performance over packing as many characters in as possible. Maybe we figure out that the width will not exceed 75% of the font-size (I'm just making this number up—who knows!). In that case, we would not allow the font-size to exceed 43px, as this is ~1.3 times our allowable width of 33 pixels per character.

Are others running into this issue frequently or is it an edge case? 

If we do find a way to solve it, would it be worth the performance drain? (Unless we are able to find a way to solve this problem with negligible performance impact.)

Ideas and feedback welcome!

**Comment by jdempcy** [comment]
If we set "word-break: break-all" in the CSS, then the text wraps as follows:
![screen shot 2016-01-26 at 4 58 01 pm](https://cloud.githubusercontent.com/assets/95693/12600356/05a10410-c44e-11e5-840a-1c2cb0596ae2.png)

This seems dispreferable to enforcing a maximum font-size, in my opinion, but may be a better solution than simply cropping the text.

**Comment by danielweck** [comment]
Hello @jdempcy 
I see you are using Readium's official "cloud reader" deployment at:
https://readium.firebaseapp.com/?epub=https%3A%2F%2Freadium.firebaseapp.com%2Fepub_content%2Fmoby_dick

When I open the same HTML file ; on its own ; in the web browser:
https://readium.firebaseapp.com/epub_content/moby_dick/OPS/introduction_001.xhtml
...and increase the font size, the text overflows beyond the right edge, instead of wrapping round the visible boundary.

This visual cropping is consistent with CSS styling expectations, I don't think it is anything specific to Readium. Limiting Readium's font size / maximum value may work for a few EPUBs, but would not prevent larger text to overflow (e.g. headings / titles authored with larger font size by default). What do you think?

**Comment by jdempcy** [comment]
I agree that the visual cropping is consistent with CSS, when an element has a positioning context other than default ("static") and with overflow: hidden.

However, in a reflowable book, it seems that the text should always reflow instead of just going off the page.

I suppose a solution is that users can just set their font size smaller.

I was thinking more of the case where the book has specified a large header size (e.g. 48px) and the rest of the text is more standard, like 12px.

It would be nice if the user could increase the body text without making the headers unreadable.

Maybe it would even be sufficient to remove "h1, h2, h3" from the jQuery selector for picking which elements to have their font size changed. Changing text size would no longer modify header text but it would allow users to bump up the body text size.

I'm okay with resolving this ticket as "won't fix" if it is simply a limitation of the reading system that we will allow users to set a font-size resulting in unreadable text. I was just wishing for a maximum that ensured text is always visible.

**Comment by jdempcy** [comment]
> Limiting Readium's font size / maximum value may work for a few EPUBs, but would not prevent larger text to overflow (e.g. headings / titles authored with larger font size by default). What do you think?

I did a few tests and found that the font-size is enforced on all elements listed, including h1, h2, h3 etc and span, div... So if we set a maximum font-size in Readium we can actually ensure that text is not cropped.

In a version of SDKLauncher-iOS I have locally, I am able to set a reasonable maximum font size and what ends up happening is this:

    - Starting with 100% font size, book renders as expected
    - Let's say I increase it to 200%. At some point while increasing the text size, the headers stop getting larger. I set them at maximum 48px.
   - As I keep increasing the font-size, after the headers stop increasing in size, the body text continues to increase until it is the same size as the headers.

Without setting these maximums, by the time the body text is as large as the header, a single header may be font-size 90px and spanning multiple screens of an iPhone or other small-screen device.

I'm not saying we need to enforce the maximums, but in my quick tests I have found that it *is* quite trivial to ensure that in the vast majority of cases, the header text does not become cropped as the body text is increased.

The primary use case I am thinking of is some books where there is a 36px header and 12px body text. I can imagine some people needing to read 20px or 30px text, and it is annoying having many of the headers cropped at 200% or 300% font-size.

Although, I don't feel so strongly about this... Feel free to close this ticket if it is something it seems like we will never address, or leave it open if we may want to circle back to this later.
